onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -1859,60 +1853,29 @@ CUDAExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
-                                                 // in that case we should still keep the node on CUDA
-                                                 bool initializer_input = graph.IsConstantInitializer(def.Name(), /*check_outer_scope*/ true);
-                                                 bool input_is_on_cpu = defs_outside_cuda.count(&def) > 0;
-                                                 if ((!initializer_input && !input_is_on_cpu) ||
-                                                     (input_is_on_cpu && cuda_kernel_def->kernel_def->IsInputOnCpu(index))) {
-                                                   should_force_outside = false;
-                                                 }
-
-                                                 if (!initializer_input) {
-                                                   all_inputs_are_initializers = false;
-                                                 }
-                                                 return Status::OK();
-                                               }));
-
-      // If all the inputs are initializers, we shouldn't force it to CPU
-      if (should_force_outside && !all_inputs_are_initializers) {
-        force_outside = true;
-      }
-    }
-#endif
-    if (!force_inside && (not_supported || force_outside)) {
-      defs_outside_cuda.insert(node.OutputDefs().cbegin(), node.OutputDefs().cend());
+    if (!force_inside && not_supported) {
       if (not_supported) {
         LOGS_DEFAULT(WARNING) << "CUDA kernel not supported. Fallback to CPU execution provider for Op type: " << node.OpType() << " node name: " << node.Name();
-      } else if (force_outside) {
-        LOGS_DEFAULT(INFO) << "Force fallback to CPU execution provider for Op type: " << node.OpType() << " node name: " << node.Name();
       }
     } else {
-      // for nodes placed on CUDA, check if its output is on CPU
-      ORT_THROW_IF_ERROR(node.ForEachWithIndex(
-          node.OutputDefs(),
-          [&](const NodeArg& def, size_t out_index) {
-            if (cuda_kernel_def->kernel_def->OutputMemoryType(out_index) != OrtMemTypeDefault)
-              defs_outside_cuda.insert(&def);
-            return Status::OK();
-          }));
-      std::unique_ptr<IndexedSubGraph> sub_graph = onnxruntime::make_unique<IndexedSubGraph>();
-      sub_graph->nodes.push_back(node.Index());
-      result.push_back(onnxruntime::make_unique<ComputeCapability>(std::move(sub_graph)));
+      candidates.push_back(node.Index());
     }
   }
+
+  // For CUDA EP, exclude the subgraph that is prefered to be placed in CPU
+  // These are usually shape related computation subgraphs
+  // Following logic can be extended for other EPs
+  std::unordered_set<NodeIndex> cpu_nodes = GetCpuPreferedNodes(graph, this, candidates);
+
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+  for (auto& node_index : candidates) {
+    if (cpu_nodes.count(node_index) > 0)
+      continue;
```

Comment:
This is nice. We might add support of compiled graph on top of the CPU preferred logic now.